Bugfix: ensure Checkbox state comparisons are correct by using Qt.CheckState(state)

psobolewskiPhD opened the pull request:
# Description

This PR fixes #5540 
Basically, the existing Checkbox comparisons fail on Qt6 because we compare the returned `state`, which is 0 or 2, with the enum Qt.CheckState.Checked. This always fails.
Now instead of using `state`, I make the comparison with `Qt.CheckState(state)`.
See pyside6 docs:
https://doc.qt.io/qtforpython/PySide6/QtCore/Qt.html#PySide6.QtCore.PySide6.QtCore.Qt.CheckState
And here's QT5 docs:
https://doc.qt.io/qt-5/qt.html#CheckState-enum

Note I also found 2 other places in the code where I fixed this, outside of labels layer checkboxes.

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

closes #5540 
closes #5556

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

psobolewskiPhD committed:
fix Qt6 checkstate

Czaki commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -403,7 +405,9 @@ def change_preserve_labels(self, state):
             Checkbox indicating if overwriting label is enabled.
         """
-        self.layer.preserve_labels = state == Qt.CheckState.Checked
+        self.layer.preserve_labels = (
+            Qt.CheckState(state) == Qt.CheckState.Checked
+        )
```
Why we use this instead of `bool(state)`? There is reason why we need to distinguish partially checked state?

psobolewskiPhD replied:
That won't work, the problem is on the right hand side.
Qt.CheckState.Checked is an enum so the comparison still fails.
This can work:
`        self.layer.contiguous = state == int(Qt.CheckState.Checked)`
but I'm not sure I like it, because it seems more likely to break again in the future.

psobolewskiPhD replied:
Actually no, this `self.layer.contiguous = state == int(Qt.CheckState.Checked)` doesn't work:
`TypeError: int() argument must be a string, a bytes-like object or a real number`

psobolewskiPhD replied:
Maybe: `self.layer.contiguous = state == Qt.CheckState.Checked.value` would work?
But still I think I prefer the more explicit comparison. 🤷‍♂️

andy-sweet replied:
I also prefer the more explicit comparison (mostly because I don't need to go look up the enum values to understand the code). But I wouldn't oppose the cast to `bool` suggested because Qt's check state is quite well known - wouldn't `self.layer.preserve_labels = bool(state)` work?

Czaki reviewed (changes requested):
As we can see changed lines are not covered by tests. Could you add tests to prevent these problems in the future?

psobolewskiPhD committed:
Tests for layer control checkboxes

psobolewskiPhD committed:
Merge branch 'qt6_checkstate_tests' into bugfix/qt6_checkstate
Add tests for the checkbox states
The tests pass on main, Qt5, but fail with main, Qt6

andy-sweet commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -369,7 +371,7 @@ def change_contig(self, state):
         state : QCheckBox
             Checkbox indicating if labels are contiguous.
```
Optional: not changed here, but the parameter's docstring is related and quite wrong. Here's a hopefully better alternative, but feel free to write your own (or even consider removing the current one): 

```suggestion
        state : int
            Integer value of Qt.CheckState that indicates the check state of contigCheckBox
```

andy-sweet commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -403,7 +405,9 @@ def change_preserve_labels(self, state):
         state : QCheckBox
```
Optional: similar to https://github.com/napari/napari/pull/5541/files#r1100890632, this docstring should probably be updated.

andy-sweet reviewed (approved):
Thanks for following up with further fixes to some of my recent changes @psobolewskiPhD . Looks good to me now, but I'll let @Czaki have the final say.

andy-sweet commented on napari/_qt/layer_controls/_tests/test_qt_labels_layer.py:
```diff
@@ -63,3 +63,22 @@ def test_changing_colormap_updates_colorbox(qtbot):
     )
+
+
+def test_layercontrols_checkboxes(qtbot):
```
Optional: slight preference for splitting this up into multiple tests and refactoring the setup code to be shared and maybe a fixture. But this is also good enough for me right now too.

psobolewskiPhD replied:
Thanks for this comment! With the help of ChatGPT I learned about fixtures and arguments and implemented this. I hope it's correct though 😅

psobolewskiPhD committed:
Merge branch 'main' into bugfix/qt6_checkstate

Czaki commented on napari/_qt/layer_controls/_tests/test_qt_labels_layer.py:
```diff
@@ -63,3 +63,22 @@ def test_changing_colormap_updates_colorbox(qtbot):
+    assert layer.show_selected_label
+    qtctrl.selectedColorCheckbox.setChecked(False)
+    assert not layer.show_selected_label
+    qtctrl.contigCheckBox.setChecked(True)
+    assert layer.contiguous
+    qtctrl.contigCheckBox.setChecked(False)
+    assert not layer.contiguous
```
Based on the previous error I prefer to add `assert not layer.contiguous` before this, or add these two lines after. 
```python
    qtctrl.contigCheckBox.setChecked(True)
    assert layer.contiguous
```

To check if `setChecked(True)` propagates properly.

Czaki commented on napari/_qt/layer_controls/_tests/test_qt_shapes_layer.py:
```diff
@@ -33,3 +33,13 @@ def test_shape_controls_edge_color(qtbot):
+    layer = Shapes(_SHAPES)
+    qtctrl = QtShapesControls(layer)
+    qtbot.addWidget(qtctrl)
+    qtctrl.textDispCheckBox.setChecked(False)
+    assert not layer.text.visible
+    qtctrl.textDispCheckBox.setChecked(True)
+    assert layer.text.visible
```
same as above

psobolewskiPhD committed:
Refactor labels_layer to use a fixture @andy-sweet

psobolewskiPhD committed:
Tweak test per @Czaki

psobolewskiPhD committed:
Refactor test_qt_labels_layer.py to use a fixture per @andy-sweet
with help of ChatGPT.
Tweak all tests per @Czaki to make sure the state propagates.

Merge branch 'qt6_checkstate_tests' into bugfix/qt6_checkstate

psobolewskiPhD committed:
Update contiguous docstring

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

psobolewskiPhD committed:
Fix qt_labels_controls.py docstrings  @andy-sweet

psobolewskiPhD committed:
fix qt_shapes_controls.py docstrings

psobolewskiPhD commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -349,7 +353,16 @@ def changeSelection(self, value):
 
     def toggle_selected_mode(self, state):
-        self.layer.show_selected_label = state == Qt.CheckState.Checked
+        """Toggle contiguous state of label layer.
```
```suggestion
        """Toggle display of selected label only.
```

psobolewskiPhD committed:
copy-paste fix

psobolewskiPhD commented on napari/_qt/layer_controls/qt_vectors_controls.py:
```diff
@@ -207,7 +207,9 @@ def change_out_of_slice(self, state):
         state : QCheckBox
             Checkbox to indicate whether to render out of slice.
         """
```
```suggestion
        state : int
             Integer value of Qt.CheckState that indicates the check state of outOfSliceCheckBox
        """
```

psobolewskiPhD committed:
fix vectors layer checkbox docstring

psobolewskiPhD committed:
fix visibility state CheckState

andy-sweet reviewed (approved):
Thanks for the extra catch.

psobolewskiPhD committed:
Merge branch 'main' into bugfix/qt6_checkstate
